core/src/main/java/com/linecorp/armeria/client/cookie/DefaultCookieJar.java

```diff
@@ -0,0 +1,218 @@
+    private final ReentrantLock lock;
+
+    DefaultCookieJar() {
+        this(CookiePolicy.acceptOriginOnly());
+    }
+
+    DefaultCookieJar(CookiePolicy cookiePolicy) {
+        this.cookiePolicy = cookiePolicy;
+        store = new Object2LongOpenHashMap<>();
+        filter = new HashMap<>();
+        lock = new ReentrantLock();
+    }
+
+    @Override
+    public Cookies get(URI uri) {
+        requireNonNull(uri, "uri");
+        if (store.isEmpty()) {
+            return Cookies.of();
+        }
+        final String host = uri.getHost();
+        final String path = uri.getPath().isEmpty() ? "/" : uri.getPath();
+        final boolean secure = isSecure(uri.getScheme());
+        final Set<Cookie> cookies = new HashSet<>();
+        lock.lock();
+        try {
+            filterGet(cookies, host, path, secure);
+        } finally {
+            lock.unlock();
+        }
+        return Cookies.of(cookies);
+    }
+
+    private boolean isSecure(String scheme) {
+        final SessionProtocol parsedProtocol;
+        if (scheme.indexOf('+') >= 0) {
+            final Scheme parsedScheme = Scheme.tryParse(scheme);
+            parsedProtocol = parsedScheme != null ? parsedScheme.sessionProtocol() : null;
+        } else {
+            parsedProtocol = SessionProtocol.find(scheme);
+        }
+        return parsedProtocol != null && parsedProtocol.isTls();
+    }
+
+    @Override
+    public void set(URI uri, Iterable<? extends Cookie> cookies, long createdTimeMillis) {
+        requireNonNull(uri, "uri");
+        requireNonNull(cookies, "cookies");
+        lock.lock();
+        try {
+            for (Cookie cookie : cookies) {
+                cookie = ensureDomainAndPath(cookie, uri);
+                // remove similar cookie if present
+                store.removeLong(cookie);
+                if ((cookie.maxAge() == Cookie.UNDEFINED_MAX_AGE || cookie.maxAge() > 0) &&
+                    cookiePolicy.accept(uri, cookie)) {
+                    store.put(cookie, createdTimeMillis);
+                    final Set<Cookie> cookieSet = filter.computeIfAbsent(cookie.domain(), s -> new HashSet<>());
+                    // remove similar cookie if present
+                    cookieSet.remove(cookie);
+                    cookieSet.add(cookie);
+                }
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    @Override
+    public CookieState state(Cookie cookie, long currentTimeMillis) {
+        requireNonNull(cookie, "cookie");
+        lock.lock();
+        try {
+            final long createdTimeMillis = store.getOrDefault(cookie, Long.MIN_VALUE);
+            if (createdTimeMillis == Long.MIN_VALUE) {
+                return CookieState.NON_EXISTENT;
+            }
+            return isExpired(cookie, createdTimeMillis, currentTimeMillis) ?
+                   CookieState.EXPIRED : CookieState.EXISTENT;
+        } finally {
+            lock.unlock();
+        }
```

Comment:
Could we reduce the lock block? I think we can apply it for `store.getOrDefault(cookie, Long.MIN_VALUE)`